### PR TITLE
Add sharding support to MongoDBContainer

### PR DESCRIPTION
**Pull request by evanchooly**
Adds sharding support but in its current form may not quite be The Testcontainers Way just yet.  Happy to work with whomever to rearrange it to its correct form.

### CODE REVIEW

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (eddumelendez) [inline modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:117]: Use `assertThat` from `Assertions` (assertj) please

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (eddumelendez) [inline modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:41]: no need to add Assertions here

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:23]: Let's not update the default version due to backward compatibility. If `new MongoDBContainer()` is used then the default version is used. Nowadays, is highly recommend to use `new MongoDBContainer("mongo:6")` instead.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:73]: Let's copy `sharding.sh` only when `shardingEnabled` is `true`

```suggestion
        if (this.shardingEnabled) {
            copyFileToContainer(MountableFile.forClasspathResource("/sharding.sh", 0777), STARTER_SCRIPT);
        }
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:42]: The following method can be added in order to opt-in for Sharding

```java
public MongoDBContainer withSharding() {
    this.shardingEnabled = true;
    return self();
}
```

```suggestion
        this(DEFAULT_IMAGE_NAME.withTag(DEFAULT_TAG));
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:46]:
```suggestion
        this(DockerImageName.parse(dockerImageName));
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:53]:
```suggestion
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:55]:
```suggestion
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:69]:
```suggestion
    }
    
    @Override
    protected void configure() {
        if (this.shardingEnabled) {
            withCreateContainerCmdModifier(cmd -> {
                cmd.withEntrypoint("sh");
            });
            withCommand("-c", "while [ ! -f " + STARTER_SCRIPT + " ]; do sleep 0.1; done; " + STARTER_SCRIPT);
            waitingFor(Wait.forLogMessage("(?i).*mongos ready.*", 1));
        } else {
            withCommand("--replSet", "docker-rs");
            waitingFor(Wait.forLogMessage("(?i).*waiting for connections.*", 1));
        }
    }
```

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:35]:
```suggestion
    private boolean shardingEnabled;
```

**Comment by evanchooly** [comment]
> Thanks for your contribution @evanchooly ! I've left some comments. I already have those changes locally but can not push them to your branch (probably because is the main branch thinking )

yeah, i noticed that after i pushed.  a brainfart.  should've done it on a branch.  maybe with these edits, i'll move it off to a branch and recreate.  but maybe it's not worth the effort.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (evanchooly) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:23]: I'm not terribly tied to this.  I won't presume to make product level decisions in my first PR.  I can change to "6" if you'd like or revert it to the 4.x version.  But 4.x is *really* old at this point so making it the default seems ... unproductive.  Your call, though.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (evanchooly) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:42]: I had it this wait to start but the way the contructor currently works, it never sees that field.  Now that I have a working form and understand the lifecycle a bit more, I think I know how to disable the default script the container runs at start up and then run the sharding script in `containerStarted()`.

In fact, when looking back over this PR (in reviewer mode vs author mode) I realized this current set up may actually end up with a non-replSet config.  I'll confirm as I rework that constructor.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (evanchooly) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:69]: +1  this is exactly what I was looking for.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:42]: There is another comment below which adds `configure` method. That way, the proper wait strategy is executed.

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (evanchooly) [reply modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:41]: i think this was for my debugging actually but it didn't leap out at me when I was cleaning stuff up.  will remove.

**Comment by evanchooly** [comment]
i think i have all the requested changes made.

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (eddumelendez) [inline modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:18]: unused import. Remove, please

**Review by eddumelendez — COMMENTED** [review]
Check fails because of an unused import. Also, please run `./gradlew :mongodb:check --no-daemon`

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (evanchooly) [reply modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:18]: Doh. In my rush to get out the door I forgot to run those checks.  Will do when I get home.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java]: @evanchooly can you revert the default tag to `4.0.10`, please? We know is very old and it belongs to the old-way to do it in Testcontainers. However, we are promoting a lot using `new MongoDBContainer("mongo:6")` :)

**Comment by evanchooly** [comment]
I wasn't clear from your comments earlier if you wanted `mongo:6` to be the default or `mongo:4.0.4`.  I just changed it back, though.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:23]: Can we revert this, please?

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (eddumelendez) [inline modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:16]: Let's keep the static import instead.

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (eddumelendez) [inline modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:27]: Let's revert this as it will break our docs, please

**Review by eddumelendez — COMMENTED** [review]
@evanchooly thanks! Just last comments before to merge it. Sorry about this, I'd applied them but do not have access.

File: modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java
Comment (evanchooly) [reply modules/mongodb/src/test/java/org/testcontainers/containers/MongoDBContainerTest.java:16]: I don't how/why this change got in there.  I prefer the static imports so this change is the opposite of what I'd consciously choose.  I wonder if IDEA "helped" me out here a little.

File: modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java
Comment (eddumelendez) [inline modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java:54]:
```suggestion
```

Comment (eddumelendez) [reply modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java]: configure is part of the Testcontainers lifecycle. So, there is no need to call it directly.

**Review by eddumelendez — APPROVED** [review]

**Comment by eddumelendez** [comment]
Thank you so much, @evanchooly ! For your contribution and patience :) This is now merge in `main` branch and it will be part of the next release.

**Comment by evanchooly** [comment]
yay!  next time i'll use a branch.  ;)

**Pull request merged**
